src/libinjector/linux/methods/linux_execve.c

```diff
@@ -0,0 +1,292 @@
+            }
+
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+
+            break;
+        }
+        case STEP2: // forks the process
+        {
+            if (!call_mmap_syscall_cb(injector, info->regs))
+                return cleanup(injector, info->regs);
+
+            char* proc_name = g_malloc0(strlen(info->proc_data.name));
+            strcpy(proc_name, info->proc_data.name);
+
+            // set the ppid and name info for the child process
+            // ( to be used in check_userspace_int3_trap )
+            injector->child_data.name = proc_name;
+            injector->child_data.ppid = info->proc_data.pid;
+
+            PRINT_DEBUG("vForking the process\n");
+
+            info->regs->rax = sys_vfork;
+            info->regs->rip = injector->syscall_addr;
+
+            // this will loosen the checks in check_userspace_int3_trap
+            // till we get the pid of the child in the next step
+            injector->fork = true;
+
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+            break;
+        }
+        case STEP3: // get child pid and runs execve
+        {
+            if (is_syscall_error(info->regs->rax, "vfork syscall failed"))
+                return cleanup(injector, info->regs);
+
+            injector->child_data.pid = info->proc_data.pid;
+            injector->child_data.tid = info->proc_data.tid;
+
+            PRINT_DEBUG("Child process pid: %d\n", injector->child_data.pid);
+
+            if (!setup_execve_syscall(injector, info->regs, injector->virtual_memory_addr))
+                return cleanup(injector, info->regs);
+
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+            break;
+        }
+        case STEP4: // handles execve and restores parent
+        {
+            if (is_child_process(injector, info))
+            {
+                is_syscall_error(info->regs->rax, "execve syscall failed");
+
+                PRINT_DEBUG("Exiting child process\n");
+                if (!setup_exit_syscall(injector, info->regs, 0))
+                {
+                    fprintf(stderr, "Fatal error: Could not cleanup properly\n");
+                    drakvuf_interrupt(drakvuf, SIGINT);
+                    return VMI_EVENT_RESPONSE_NONE;
+                }
+
+                // this is being done so that the parent can also be cleared
+                return override_step(injector, STEP4, VMI_EVENT_RESPONSE_SET_REGISTERS);
+            }
+            else
+            {
+                PRINT_DEBUG("Restoring parent registers\n");
+                memcpy(info->regs, &injector->saved_regs, sizeof(x86_registers_t));
+
+                // free the post_syscall_trap
+                free_bp_trap(drakvuf, injector, info->trap);
+            }
+
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+            break;
+
+        }
+        case STEP5: // exit drakvuf loop
+        {
+            if (is_child_process(injector, info))
```

Comment:
I think this should never happen. `exit` should not fail.

Reply:
Yeah, I know that but it is just a fail-safe. If something ever happens like that then we should know about it. The cleanup in this seems to be somewhat tricky and I would really like this to remain there

Reply:
But leaving this is code suggest that we expect this. I would even prefer `assert`.

Reply:
Btw, do you think that we should use `assert` only using debug build? ( As if an assertion occurs, the code exits very abruptly causing the VM to get in an inconsistent state as `close_vmi` and everything doesn't happen )

Reply:
@kscieslinski ping

Reply:
IMO if we get into an unexpected state, we don't really have to clean up after ourselves since the VM is somehow broken already. I think that we could even call `abort()`. What do you think?

Reply:
In this state, only a child process will remain in the VM which won't be killed. I don't think that it should make the VM unusable. The process can be killed manually afterwards, but if `close_vmi` doesn't occur when we hit assertion, then the VM stays in a paused state making it unusable

Reply:
Ok, then just make sure to add some log like `fprintf(stderr, "Should never happen...")`.

Reply:
done